### PR TITLE
Don't fail on insufficient parameters in `ParameterFormatter` (#2337)

### PR DESCRIPTION
**Pull request by vy**

### CODE REVIEW

File: log4j-api-test/src/test/java/org/apache/logging/log4j/message/ParameterFormatterTest.java [hunk 1]
```diff
@@ -17,21 +17,27 @@
...
 import org.junit.jupiter.params.provider.CsvSource;
 import org.junit.jupiter.params.provider.MethodSource;
+import org.mockito.ArgumentCaptor;
 
 /**
  * Tests {@link ParameterFormatter}.
  */
-public class ParameterFormatterTest {
+@UsingStatusLoggerMock
+class ParameterFormatterTest {
```
Comment (ppkarwasz) [inline log4j-api-test/src/test/java/org/apache/logging/log4j/message/ParameterFormatterTest.java:40]: I am not a big fan of mocking the entire `StatusLogger`: this prevents the test to run in parallel with other tests.

I prefer the [`@UsingStatusListener`](https://github.com/apache/logging-log4j2/blob/2.x/log4j-api-test/src/main/java/org/apache/logging/log4j/test/junit/UsingStatusListener.java) extension to register a per-test status listener.

Of course neither of us is impartial in the choice of JUnit extension to use. ;-)

File: log4j-api/src/main/java/org/apache/logging/log4j/message/ParameterFormatter.java [hunk 2]
```diff
@@ -236,7 +240,10 @@ static void formatMessage(
             final String message = String.format(
                     "found %d argument placeholders, but provided %d for pattern `%s`",
                     analysis.placeholderCount, args.length, pattern);
-            throw new IllegalArgumentException(message);
```
Comment (ppkarwasz) [inline log4j-api/src/main/java/org/apache/logging/log4j/message/ParameterFormatter.java:239]: Personally I am pretty much in favor of throwing an exception here: [`java.util.Formatter`](https://docs.oracle.com/javase/8/docs/api/java/util/Formatter.html) does it, so we can also throw.

On the other hand we should probably catch the exception in `ParameterizedMessage#getFormattedMessage`, as it happens in other kinds of messages:

https://github.com/apache/logging-log4j2/blob/28b70d1d7fe63417e8eaa1abe8bf992363060515/log4j-api/src/main/java/org/apache/logging/log4j/message/StringFormattedMessage.java#L120-L126

https://github.com/apache/logging-log4j2/blob/28b70d1d7fe63417e8eaa1abe8bf992363060515/log4j-api/src/main/java/org/apache/logging/log4j/message/MessageFormatMessage.java#L113-L119

**Review by ppkarwasz — CHANGES_REQUESTED** [review]
Personally I would maintain the current logic that throws exceptions, I would just catch them in the call sites.

I looked in Log4j 2.19.0 how formatting was done. Using the `StringFormatterMessageFactory` a call to `logger.info("Hello %s and %s!", "world")` gives:

```
2024-03-04 11:18:57,550 main ERROR Unable to format msg: Hello %s and %s!
java.util.MissingFormatArgumentException: Format specifier '%s'
Hello %s and %s!
```

whereas the default factories for a call to `logger.info("Hello {} and {}!", "world")` give just:
```
Hello world and {}!
```

I don't think the behavior of Log4j 2.19.0 is correct. I would expect rather:
```
2024-03-04 11:18:57,550 main ERROR Unable to format msg: Hello {} and {}!
java.lang.IllegalArgumentException: Found 2 argument placeholders, but only 1 was provided.
Hello {} and {}!
```

File: log4j-api/src/main/java/org/apache/logging/log4j/message/ParameterFormatter.java (hunk 2 above)
Comment (vy) [reply log4j-api/src/main/java/org/apache/logging/log4j/message/ParameterFormatter.java]: I have updated it as you requested. Could you review the changes in 1084855467f554f2dabf34af4d860018fb098fd8, please?

File: log4j-api/src/main/java/org/apache/logging/log4j/message/ParameterizedMessage.java [hunk 3]
```diff
@@ -274,7 +278,12 @@ public void formatTo(final StringBuilder buffer) {
             buffer.append(formattedMessage);
         } else {
             final int argCount = args != null ? args.length : 0;
-            ParameterFormatter.formatMessage(buffer, pattern, args, argCount, patternAnalysis);
+            try {
+                ParameterFormatter.formatMessage(buffer, pattern, args, argCount, patternAnalysis);
+            } catch (final Exception error) {
+                STATUS_LOGGER.error("message formatting failure", error);
```
Comment (ppkarwasz) [inline log4j-api/src/main/java/org/apache/logging/log4j/message/ParameterizedMessage.java:284]: Nit: the message is "Unable to format msg: {}" in the case of `StringFormattedMessage`. Can we use the same message everywhere?

**Review by ppkarwasz — APPROVED** [review]
LGTM

File: log4j-api/src/main/java/org/apache/logging/log4j/message/ParameterizedMessage.java (hunk 3 above)
Comment (vy) [reply log4j-api/src/main/java/org/apache/logging/log4j/message/ParameterizedMessage.java]: Fixed in a40714e415813433df099f6d15de5b511ea3f910.

File: log4j-api-test/src/test/java/org/apache/logging/log4j/message/ParameterFormatterTest.java (hunk 1 above)
Comment (vy) [reply log4j-api-test/src/test/java/org/apache/logging/log4j/message/ParameterFormatterTest.java:37]: Piotr, if I would have been a programmer who also happens to be a serial killer, you would not want me to see what you have done in `UsingStatusListener` et al. using `ThreadLocal`s. :sunglasses: I find it a liability with several fragile assumptions just to make tests needing `StatusLogger` work in parallel. I have so much dislike towards such design choices, but I will spare that for another time.

Just to show how much I like you, I have replaced `@UsingStatusLoggerMock` with `@UsingStatusListener` in 186c580d2cda1f301e7b5eab14ded8a15e5b54f5 as you requested. :stuck_out_tongue_closed_eyes:

**Review by ppkarwasz — APPROVED** [review]

**Pull request merged**